test/e2e/taint_toleration_test.go

```diff
@@ -0,0 +1,212 @@
+					// specified here are validated only on a very superficial level, i.e., the flow
+					// functions. For further evaluations, specifically the correctness check
+					// of the affinity and topology spread constraint logic, see the scheduler
+					// integration tests.
+					Affinity: &placementv1beta1.Affinity{
+						ClusterAffinity: &placementv1beta1.ClusterAffinity{
+							RequiredDuringSchedulingIgnoredDuringExecution: &placementv1beta1.ClusterSelector{
+								ClusterSelectorTerms: []placementv1beta1.ClusterSelectorTerm{
+									{
+										LabelSelector: &metav1.LabelSelector{
+											MatchLabels: map[string]string{
+												regionLabelName: regionLabelValue1,
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+					TopologySpreadConstraints: []placementv1beta1.TopologySpreadConstraint{
+						{
+							MaxSkew:           ptr.To(int32(1)),
+							TopologyKey:       envLabelName,
+							WhenUnsatisfiable: placementv1beta1.DoNotSchedule,
+						},
+					},
+					Tolerations: buildTolerations(tolerateClusterNames),
+				},
+				Strategy: placementv1beta1.RolloutStrategy{
+					Type: placementv1beta1.RollingUpdateRolloutStrategyType,
+					RollingUpdate: &placementv1beta1.RollingUpdateConfig{
+						UnavailablePeriodSeconds: ptr.To(2),
+					},
+				},
+			},
+		}
+		Expect(hubClient.Create(ctx, crp)).To(Succeed(), "Failed to create cluster resource placement")
+	})
+
+	It("should update cluster resource placement status as expected", func() {
+		statusUpdatedActual := crpStatusUpdatedActual(workResourceIdentifiers(), []string{memberCluster1EastProdName}, []string{memberCluster2EastCanaryName}, "0")
+		Eventually(statusUpdatedActual, eventuallyDuration, eventuallyInterval).Should(Succeed(), "Failed to update cluster resource placement status as expected")
+	})
+
+	It("should place resources on the selected clusters with tolerated taint", func() {
+		targetClusters := []*framework.Cluster{memberCluster1EastProd}
+		for _, cluster := range targetClusters {
+			resourcePlacedActual := workNamespaceAndConfigMapPlacedOnClusterActual(cluster)
+			Eventually(resourcePlacedActual, eventuallyDuration, eventuallyInterval).Should(Succeed(), "Failed to place resources on the selected clusters")
+		}
+	})
+
+	It("should ensure no resources exist on member clusters with untolerated taint", func() {
+		unSelectedClusters := []*framework.Cluster{memberCluster2EastCanary}
+		for _, cluster := range unSelectedClusters {
+			resourceRemovedActual := workNamespaceRemovedFromClusterActual(cluster)
+			Eventually(resourceRemovedActual, eventuallyDuration, eventuallyInterval).Should(Succeed(), "Failed to check if resources doesn't exist on member cluster")
+		}
+	})
+
+	AfterAll(func() {
+		// Remove taint from member cluster 1, 2.
+		removeTaintsFromMemberClusters(taintClusterNames)
+		ensureCRPAndRelatedResourcesDeletion(crpName, []*framework.Cluster{memberCluster2EastCanary})
+	})
+})
```

Comment:
Also, could you add a test case to verify the NoSchedule behavior? i.e., adding a taint won't trigger the removal of resources from a cluster that has been selected?

Reply:
It can be done in separate PR, of course.